python/taichi/ad/_ad.py

```diff
@@ -205,24 +206,33 @@ def __enter__(self):
 
         # Attach the context manager to runtime
         self.runtime.target_tape = self
+        return self
 
     def __exit__(self, _type, value, tb):
         self.runtime.target_tape = None
         if self.eval_on_exit:
             self.grad()
+        if self.validation:
+            for f in self.validation_mode_kernels:
+                f.autodiff_mode = AutodiffMode.NONE
 
     def insert(self, func, args):
+        if self.validation:
+            if func.autodiff_mode == AutodiffMode.VALIDATION:
```

Comment:
This still looks a bit weird to me: when did we change this to valiadation mode (logically it should happen here and you save their original modes for restoring right? ;)

Reply:
I was intended to change the mode here. However found the autodiff mode is used when creating kernel before the insert happens LOL.

Reply:
Hmmm IIUC the kernel autodiff mode before entering the tape is not always `None`? Does restoring to `None` make sense? (if so please feel free to add a comment ;)

Reply:
The kernel in validation mode always has a `None` mode before entering the tape. And only the kernels in validation mode  are recorded in the `self.validation_mode_kernels`.